// @antv/l7-three/lib/core/baseLayer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _l7Layers = require("@antv/l7-layers");

var _three = require("three");

var _threeRenderService = require("./threeRenderService");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var DEG2RAD = Math.PI / 180;

var ThreeJSLayer = function (_BaseLayer) {
  (0, _inherits2.default)(ThreeJSLayer, _BaseLayer);

  var _super = _createSuper(ThreeJSLayer);

  function ThreeJSLayer() {
    var _this;

    (0, _classCallCheck2.default)(this, ThreeJSLayer);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "type", 'custom');
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "threeRenderService", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "isUpdate", false);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "update", null);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "scene", new _three.Scene());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderer", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "animateMixer", []);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "center", void 0);
    return _this;
  }

  (0, _createClass2.default)(ThreeJSLayer, [{
    key: "setUpdate",
    value: function setUpdate(callback) {
      this.update = callback;
      this.isUpdate = true;
    }
  }, {
    key: "getModelMatrix",
    value: function getModelMatrix(lnglat) {
      var altitude = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
      var rotation = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [0, 0, 0];
      var scale = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : [1, 1, 1];
      return new _three.Matrix4().fromArray(this.mapService.getModelMatrix(lnglat, altitude, rotation, scale, this.threeRenderService.center));
    }
  }, {
    key: "getTranslateMatrix",
    value: function getTranslateMatrix(lnglat) {
      var altitude = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
      return this.getModelMatrix(lnglat, altitude, [0, 0, 0], [1, 1, 1]);
    }
  }, {
    key: "applyObjectLngLat",
    value: function applyObjectLngLat(object, lnglat) {
      var altitude = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
      var positionMatrix = this.getTranslateMatrix(lnglat, altitude);
      object.applyMatrix4(positionMatrix);
    }
  }, {
    key: "setObjectLngLat",
    value: function setObjectLngLat(object, lnglat) {
      var altitude = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;

      var _this$lnglatToCoord = this.lnglatToCoord(lnglat),
          _this$lnglatToCoord2 = (0, _slicedToArray2.default)(_this$lnglatToCoord, 2),
          x = _this$lnglatToCoord2[0],
          y = _this$lnglatToCoord2[1];

      object.position.set(x, y, altitude);
    }
  }, {
    key: "lnglatToCoord",
    value: function lnglatToCoord(lnglat) {
      var _this$mapService;

      var _this$mapService$lngL = (_this$mapService = this.mapService) === null || _this$mapService === void 0 ? void 0 : _this$mapService.lngLatToCoord(lnglat, this.threeRenderService.center),
          _this$mapService$lngL2 = (0, _slicedToArray2.default)(_this$mapService$lngL, 2),
          x = _this$mapService$lngL2[0],
          y = _this$mapService$lngL2[1];

      return [x, y];
    }
  }, {
    key: "getObjectLngLat",
    value: function getObjectLngLat(object) {
      var coord = [object.position.x, object.position.y];
      return [0, 0];
    }
  }, {
    key: "adjustMeshToMap",
    value: function adjustMeshToMap(object) {
      object.up = new _three.Vector3(0, 0, 1);
      var defaultLngLat = this.mapService.getCenter();
      var modelMatrix = this.getModelMatrix([defaultLngLat.lng, defaultLngLat.lat], 0, [Math.PI / 2, -Math.PI, 0], [1, 1, 1]);
      object.applyMatrix4(modelMatrix);
    }
  }, {
    key: "setMeshScale",
    value: function setMeshScale(object) {
      var x = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1;
      var y = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 1;
      var z = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 1;
      var scaleMatrix = new _three.Matrix4();
      scaleMatrix.scale(new _three.Vector3(x, y, z));
      object.applyMatrix4(scaleMatrix);
    }
  }, {
    key: "buildModels",
    value: function buildModels() {
      this.threeRenderService = this.getContainer().get(_threeRenderService.ThreeRenderServiceType);
      var config = this.getLayerConfig();

      if (config && config.onAddMeshes) {
        config.onAddMeshes(this.scene, this);
      }
    }
  }, {
    key: "renderModels",
    value: function renderModels() {
      var _this2 = this;

      if (this.isUpdate && this.update) {
        this.update();
      }

      var gl = this.rendererService.getGLContext();
      this.rendererService.setCustomLayerDefaults();
      var cullFace = this.mapService.constructor.name === 'AMapService' ? gl.BACK : gl.FRONT;
      gl.cullFace(cullFace);
      var renderer = this.threeRenderService.renderer;
      renderer.state.reset();
      renderer.autoClear = false;
      var camera = this.threeRenderService.getRenderCamera();
      renderer.render(this.scene, camera);
      this.rendererService.setState();
      this.animateMixer.forEach(function (mixer) {
        mixer.update(_this2.getTime());
      });
      this.rendererService.setState();
      this.rendererService.setDirty(true);
      return this;
    }
  }, {
    key: "renderAMapModels",
    value: function renderAMapModels() {
      var _this3 = this;

      var gl = this.rendererService.getGLContext();
      this.rendererService.setCustomLayerDefaults();
      var renderer = this.threeRenderService.renderer;
      renderer.state.reset();
      renderer.autoClear = false;
      renderer.render(this.scene, this.threeRenderService.getRenderCamera());
      this.animateMixer.forEach(function (mixer) {
        mixer.update(_this3.getTime());
      });
      this.rendererService.setBaseState();
      this.rendererService.setDirty(true);
      return this;
    }
  }, {
    key: "getRenderCamera",
    value: function getRenderCamera() {
      return this.threeRenderService.getRenderCamera();
    }
  }, {
    key: "addAnimateMixer",
    value: function addAnimateMixer(mixer) {
      this.animateMixer.push(mixer);
    }
  }, {
    key: "setBottomColor",
    value: function setBottomColor(color) {
      console.warn('empty function');
    }
  }, {
    key: "getBottomColor",
    value: function getBottomColor() {
      return 'rgba(0, 0, 0, 0)';
    }
  }]);
  return ThreeJSLayer;
}(_l7Layers.BaseLayer);

exports.default = ThreeJSLayer;
